lib/webauthn/metadata/client.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require "jwt"
+require "net/http"
+
+module WebAuthn
+  module Metadata
+    class Client
+      class DataIntegrityError < StandardError; end
+      class InvalidHashError < DataIntegrityError; end
+      class UnverifiedSigningKeyError < DataIntegrityError; end
+
+      DEFAULT_HEADERS = {
+        "Content-Type" => "application/json",
+        "User-Agent" => "webauthn/#{WebAuthn::VERSION} (Ruby)"
+      }.freeze
+
+      def self.fido_trust_store
+        store = OpenSSL::X509::Store.new
+        store.purpose = OpenSSL::X509::PURPOSE_ANY
+        store.flags = OpenSSL::X509::V_FLAG_CRL_CHECK | OpenSSL::X509::V_FLAG_CRL_CHECK_ALL
+        file = File.read(File.join(__dir__, "Root.cer"))
+        store.add_cert(OpenSSL::X509::Certificate.new(file))
+      end
+
+      def initialize(token)
+        @token = token
+      end
+
+      def download_toc(uri, trust_store: self.class.fido_trust_store)
+        response = get_with_token(uri)
+        payload, _ = JWT.decode(response, nil, true, algorithms: ["ES256"]) do |headers|
+          verified_public_key(headers["x5c"], trust_store)
+        end
+        payload
+      end
+
+      def download_entry(uri, expected_hash:)
+        response = get_with_token(uri)
+        unless SecureCompare.compare(OpenSSL::Digest::SHA256.digest(response), Base64.urlsafe_decode64(expected_hash))
+          raise(InvalidHashError)
+        end
+
+        decoded_body = Base64.urlsafe_decode64(response)
+        JSON.parse(decoded_body)
+      end
+
+      private
+
+      def get_with_token(uri)
+        if @token && !@token.empty?
+          uri.path += "/" unless uri.path.end_with?("/")
```

Comment:
mmm I'm not quite getting the `/` at the end, is it instead `?` or maybe I'm missing something?
I'm pretty sure it's me bc else not a single request done for this would have worked 😄

Reply:
👍 thank you!